pkg/distributor/query.go

```diff
@@ -318,6 +324,10 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 	reqStats.AddFetchedChunkBytes(uint64(resp.ChunksSize()))
 	reqStats.AddFetchedChunks(uint64(resp.ChunksCount()))
 
+	// Reuse ingester query stream responses.
+	for _, response := range responses {
+		ingester_client.ReuseQueryStreamResponse(response)
+	}
```

Comment:
Have you considered moving it to a `defer()` (defined above) so that we reuse responses also when this function returns an error (e.g. limits have been reached)?

Reply:
Didn't do that initially to avoid a race condition, as the reading goroutine could still be writing to the `responses`  slice when an error occurs. 

However, having a second look, I think we could simply hang on the `doneReading` channel at the begining of the `defer()` function to avoid this type of situation.

Changing it. 👍